File: backend/internal/api/points.go
package api

import (
	"context"
	"net/http"

	"github.com/gin-gonic/gin"

	"forklift-training/internal/middleware"
	"forklift-training/internal/service"
	"forklift-training/pkg/response"
)

// pointsErrStatus 积分域哨兵→状态码表（#610，ADR-0024）：已领取/额度/余额不足/已兑换等
// 业务冲突 → 400，不存在类 → 404；未命中兜底 400——积分域 service 错误均为业务错误，不以 500 暴露。
var pointsErrStatus = &errStatusTable{
	entries: []errStatusEntry{
		{service.ErrTaskNotFound, http.StatusNotFound},
		{service.ErrCourseNotFound, http.StatusBadRequest},
		{service.ErrCourseNotRedeemable, http.StatusBadRequest},
		{service.ErrAlreadyClaimed, http.StatusBadRequest},
		{service.ErrDailyClaimLimit, http.StatusBadRequest},
		{service.ErrInsufficientPoints, http.StatusBadRequest},
		{service.ErrAlreadyRedeemed, http.StatusBadRequest},
	},
	fallback: http.StatusBadRequest,
}

// PointsHandler 积分 handler
type PointsHandler struct {
	svc *service.PointsService
}

func NewPointsHandler(svc *service.PointsService) *PointsHandler {
	return &PointsHandler{svc: svc}
}

// RegisterPointsRoutes 注册 /api/points 蓝图（需登录，hrwai_user）
func RegisterPointsRoutes(rg *gin.RouterGroup, rd RouterDeps, svc *service.PointsService) {
	h := NewPointsHandler(svc)
	g := rg.Group("/points", middleware.JWTAuth(rd.Session), middleware.RoleRequired("hrwai_user"))
	g.GET("/balance", h.GetBalance)
	g.GET("/ledger", h.GetLedger)
	g.GET("/tasks", h.GetTasks)
	g.POST("/tasks/:code/claim", h.Claim)
	g.POST("/shop/course/:courseId/redeem", h.RedeemCourse)
	g.POST("/shop/:sku/redeem", h.RedeemShop)
}

// GetBalance 获取余额
// GetBalance 积分余额 GET /api/points/balance
// @Summary 积分余额
// @Description 学员查看自己的积分余额
// @Tags 学员端-积分
// @Produce json
// @Security BearerAuth
// @Success 200 {object} response.R "余额"
// @Failure 401 {object} response.R "未认证"
// @Router /points/balance [get]
func (h *PointsHandler) GetBalance(c *gin.Context) {
	Endpoint[struct{}, service.PointsBalanceResult]{
		Invoke: func(ctx context.Context, _ *struct{}) (*service.PointsBalanceResult, error) {
			return h.svc.GetBalance(middleware.CurrentUserID(c))
		},
	}.Handle(c)
}

// GetLedger 流水
// GetLedger 积分流水 GET /api/points/ledger
// @Summary 积分流水
// @Description 学员查看自己的积分流水（分页）
// @Tags 学员端-积分
// @Produce json
// @Security BearerAuth
// @Param page query int false "页码"
// @Param page_size query int false "每页数量"
// @Success 200 {object} response.R "流水"
// @Failure 401 {object} response.R "未认证"
// @Router /points/ledger [get]
func (h *PointsHandler) GetLedger(c *gin.Context) {
	Endpoint[struct{}, service.PointsLedgerResult]{
		Parse: func(c *gin.Context) (*struct{}, error) { return &struct{}{}, nil },
		Invoke: func(ctx context.Context, _ *struct{}) (*service.PointsLedgerResult, error) {
			page := atoiDefault(c.Query("page"), 1)
			pageSize := atoiDefault(c.Query("page_size"), 20)
			// #512：direction 收支方向筛选（"" 全部 / "in" 收入 / "out" 支出）
			direction := c.Query("direction")
			return h.svc.GetLedgerFiltered(middleware.CurrentUserID(c), page, pageSize, "", direction)
		},
	}.Handle(c)
}

// GetTasks 任务列表
// GetTasks 积分任务中心 GET /api/points/tasks
// @Summary 积分任务中心
// @Description 学员查看积分任务列表（todo/claimable/claimed 三态）
// @Tags 学员端-积分
// @Produce json
// @Security BearerAuth
// @Success 200 {object} response.R "任务列表"
// @Failure 401 {object} response.R "未认证"
// @Router /points/tasks [get]
func (h *PointsHandler) GetTasks(c *gin.Context) {
	Endpoint[struct{}, service.PointsTasksResult]{
		Invoke: func(ctx context.Context, _ *struct{}) (*service.PointsTasksResult, error) {
			return h.svc.GetTasks(middleware.CurrentUserID(c))
		},
	}.Handle(c)
}

// Claim 领取
// Claim 领取积分任务 POST /api/points/tasks/:code/claim
// @Summary 领取任务积分
// @Description 学员领取积分任务奖励（防双领 + 幂等）
// @Tags 学员端-积分
// @Produce json
// @Security BearerAuth
// @Param code path string true "任务编码"
// @Success 200 {object} response.R "已领取"
// @Failure 400 {object} response.R "已领取/额度不足"
// @Failure 401 {object} response.R "未认证"
// @Router /points/tasks/{code}/claim [post]
func (h *PointsHandler) Claim(c *gin.Context) {
	code := c.Param("code")
	if code == "" {
		response.BadRequest(c, "任务 code 不能为空")
		return
	}
	Endpoint[struct{}, service.PointsClaimResult]{
		Invoke: func(ctx context.Context, _ *struct{}) (*service.PointsClaimResult, error) {
			return h.svc.Claim(ctx, middleware.CurrentUserID(c), code)
		},
		// #610：哨兵→状态码收编至 pointsErrStatus（已领取类 400、任务不存在 404），文案零漂移
		ErrStatus: pointsErrStatus,
	}.Handle(c)
}

// RedeemCourse 兑换课程
// RedeemCourse 兑换课程 POST /api/points/shop/course/:courseId/redeem
// @Summary 兑换课程
// @Description 学员用积分兑换课程（锁 + 已拥有校验 + 幂等）
// @Tags 学员端-积分
// @Produce json
// @Security BearerAuth
// @Param courseId path int true "课程 ID"
// @Success 200 {object} response.R "兑换成功"
// @Failure 400 {object} response.R "余额不足/已拥有"
// @Failure 401 {object} response.R "未认证"
// @Router /points/shop/course/{courseId}/redeem [post]
func (h *PointsHandler) RedeemCourse(c *gin.Context) {
	courseID, err := pathInt(c, "courseId", "课程ID无效")
	if err != nil {
		response.BadRequest(c, err.Error())
		return
	}
	Endpoint[struct{}, service.RedeemResult]{
		Invoke: func(ctx context.Context, _ *struct{}) (*service.RedeemResult, error) {
			return h.svc.RedeemCourse(ctx, middleware.CurrentUserID(c), courseID)
		},
		// #610：哨兵→状态码收编至 pointsErrStatus（余额不足/已拥有等 → 400），文案零漂移
		ErrStatus: pointsErrStatus,
	}.Handle(c)
}

// RedeemShop 兑换商城
// RedeemShop 兑换商城商品 POST /api/points/shop/:sku/redeem
// @Summary 兑换商城商品
// @Description 学员用积分兑换商城商品（锁 + 已拥有校验 + 幂等）
// @Tags 学员端-积分
// @Produce json
// @Security BearerAuth
// @Param sku path string true "商品 SKU"
// @Success 200 {object} response.R "兑换成功"
// @Failure 400 {object} response.R "余额不足/已拥有"
// @Failure 401 {object} response.R "未认证"
// @Router /points/shop/{sku}/redeem [post]
func (h *PointsHandler) RedeemShop(c *gin.Context) {
	sku := c.Param("sku")
	if sku == "" {
		response.BadRequest(c, "sku 不能为空")
		return
	}
	Endpoint[struct{}, service.RedeemResult]{
		Invoke: func(ctx context.Context, _ *struct{}) (*service.RedeemResult, error) {
			return h.svc.RedeemShop(ctx, middleware.CurrentUserID(c), sku)
		},
		// #610：哨兵→状态码收编至 pointsErrStatus（余额不足/已拥有等 → 400），文案零漂移
		ErrStatus: pointsErrStatus,
	}.Handle(c)
}

// helpers already in helpers.go (atoiDefault)

File: frontend/src/api/points.ts
// 已迁移模块：走 unwrappedRequest（拦截器解包信封，成功直接返回业务数据 Promise<T>，
// 业务失败抛错并统一 toast，调用方不再自检 res.code）
//
// 响应类型**不再手写**：唯一事实源是后端注解 → backend/docs/swagger.json →
// `cd backend && go run ./cmd/gen-apitypes`（ADR-0048 决策 1/3，issue #965 片七）。
// 本文件只留请求壳、端点装配与**名称适配**，入参（query / body）类型不生成、仍手写
// （ADR-0048 决策 3：swag 对 body 描述弱）。
//
// 别名规则（片一统一口径）：旧名与生成形状一致时保留旧名（`export type 旧名 = 生成名`）；
// 形状不同时删旧名、直接用生成名（差异逐条记于片七字段级差异清单）。
import { unwrappedRequest } from './request'
import type { TaskGroup, TaskStatus } from '@/utils/taskCenter'
import type {
  PointsBalanceResult,
  PointsClaimResult,
  PointsLedgerItem,
  PointsLedgerResult,
  PointsTaskItem as GeneratedPointsTaskItem,
  PointsTasksResult,
  RedeemResult
} from './generated/points'

export type {
  PointsBalanceResult as PointsBalance,
  PointsLedgerItem,
  PointsLedgerResult as PointsLedgerData,
  RedeemResult
}

/**
 * 任务三态与分组：注解层只到 `string`（swag 尚无 enum → TS 联合的渲染能力，
 * 见 ADR-0048 片一「已知限制」），故这里从生成物**收窄**而不是手写整份形状
 * —— 字段集仍由生成物唯一决定，仅 status / group 两个封闭值集在前端窄化。
 */
export type PointsTaskItem = Omit<GeneratedPointsTaskItem, 'status' | 'group'> & {
  status: TaskStatus
  group: TaskGroup
}

export type PointsTasksData = Omit<PointsTasksResult, 'tasks'> & { tasks: PointsTaskItem[] }
export type PointsClaimData = Omit<PointsClaimResult, 'task_status'> & { task_status: TaskStatus }

// 静默头（#409）：领取与任务/余额读取都不再由请求壳统一 toast——页面用自有错误态与
// 语义分级提示，避免「拦截器弹一次 + 页面弹一次」的双 toast。
const SILENT = { headers: { 'X-Silent': '1' } }

export const pointsApi = {
  getBalance() {
    return unwrappedRequest.get<PointsBalanceResult>('/points/balance', SILENT)
  },
  getLedger(params: { page?: number; page_size?: number; direction?: 'in' | 'out' }) {
    return unwrappedRequest.get<PointsLedgerResult>('/points/ledger', { params })
  },
  getTasks() {
    return unwrappedRequest.get<PointsTasksData>('/points/tasks', SILENT)
  },
  claim(code: string) {
    return unwrappedRequest.post<PointsClaimData>('/points/tasks/' + code + '/claim', undefined, SILENT)
  },
  redeemCourse(courseId: number) {
    return unwrappedRequest.post<RedeemResult>('/points/shop/course/' + courseId + '/redeem')
  },
  redeemShop(sku: string) {
    return unwrappedRequest.post<RedeemResult>('/points/shop/' + sku + '/redeem')
  },
}
